lib/tree.rb

```diff
@@ -16,40 +16,110 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+# add helper 
+  def helper_add(current, new_node)
+    return new_node if current.nil? 
+
+    if new_node.key <= current.key 
+      current.left = helper_add(current.left , new_node)
+    else   
+      current.right = helper_add(current.right, new_node)
+    end 
+    return current
+  end 
+
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def add(key, value)
-    raise NotImplementedError
+    @root = helper_add(@root, TreeNode.new(key,value))
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # helper method 
+  def find_helper(current, key)
+    return nil if current.nil? 
+
+    if current.key == key 
+      return current.value 
+    elsif current.key < key 
+      current = current.right 
+    else  
+      current = current.left 
+    end 
+    return find_helper(current,key)
+  end 
+  # Time Complexity: log(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current)
+    result = []
+    return [] if current.nil? 
+    result += inorder_helper(current.left)
+    result.push({:key=>current.key, :value => current.value})
+    result += inorder_helper(current.right)
+    return result
```

Comment:
I just want to note that the array concatenation you are doing here would add significantly to the time complexity.  It's better that you pass the array as a parameter to the helper.